libsolidity/analysis/ControlFlowGraph.h

```diff
@@ -0,0 +1,140 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/ast/ASTVisitor.h>
+#include <libsolidity/interface/ErrorReporter.h>
+
+#include <map>
+#include <memory>
+#include <stack>
+#include <vector>
+
+namespace dev
+{
+namespace solidity
+{
+
+/// Basic Control Flow Block.
+/// Basic block of control flow. Consists of a set of AST nodes
+/// for which control flow is always linear.
+struct ControlFlowBlock
+{
+	/// All variable declarations executed in this control flow block.
+	std::vector<VariableDeclaration const*> variableDeclarations;
+	/// All expressions executed in this control flow block (this includes all subexpressions!).
+	std::vector<Expression const*> expressions;
+	/// All inline assembly statements executed in this control flow block.
+	std::vector<InlineAssembly const*> inlineAssemblyStatements;
+	/// If control flow returns in this node, the return statement is stored in returnStatement,
+	/// otherwise returnStatement is nullptr.
+	Return const* returnStatement = nullptr;
+};
+
+/// Node of the Control Flow Graph.
+/// The control flow is a directed graph connecting control flow blocks.
+/// An arc between two nodes indicates that the control flow can possibly
+/// move from its start node to its end node during execution.
+struct CFGNode
+{
+	/// Entry nodes. All CFG nodes from which control flow may move into this node.
+	std::vector<CFGNode*> entries;
```

Comment:
Does it make sense to just add these as members to ControlFlowBlock?

Reply:
I intentionally divided the node, i.e. the graph structure, from its contents. This may allow further optimizations in the future. For example at the moment the modifier flow is copied entirely when "applying" the modifier to the function flow - in the future only the graph structure could be duplicated, but the contained ControlFlowBlock could be the same object.